lib/imagery/manage_signatures.c

```diff
@@ -0,0 +1,254 @@
+    }
+
+    if (type == SIGFILE_TYPE_SIG)
+        sprintf(selem, "signatures%csig", HOST_DIRSEP);
+    else if (type == SIGFILE_TYPE_SIGSET) {
+        sprintf(selem, "signatures%csigset", HOST_DIRSEP);
+    }
+
+    G_make_mapset_element(selem);
+    G_file_name(old_path, selem, old_name, smapset);
+    G_file_name(new_path, selem, tname, G_mapset());
+    if (G_copy_file(old_path, new_path) != 1) {
+        G_warning(_("Unable to copy <%s> to current mapset as <%s>"),
+                  G_fully_qualified_name(old_name, smapset), tname);
+        return 1;
+    }
+    return 0;
+}
+
+/*!
+ * \brief Rename a signature file
+ *
+ * If rename fails, prints warning messages and returns 1.
+ * It is safe to pass fully qualified names.
+ *
+ * \param type SIGFILE_TYPE_ signature type
+ * \param name of old signature
+ * \param name of new signature
+ * \return 0 on success
+ * \return 1 on failure
+ */
+int I_signatures_rename(int type, const char *old_name, const char *new_name)
+{
+    char sname[GNAME_MAX], tname[GNAME_MAX], tmapset[GMAPSET_MAX];
+    const char *smapset;
+    char selem[GNAME_MAX];
+    char old_path[GPATH_MAX], new_path[GPATH_MAX];
+
+    G_debug(1, "I_signatures_rename(%d, %s, %s);", type, old_name, new_name);
+
+    /* Rename only if source and destinaiton mapset is the current mapset */
+    if (G_name_is_fully_qualified(old_name, sname, tmapset)) {
+        if (strcmp(tmapset, G_mapset()) != 0) {
+            G_warning(_("%s is not in the current mapset (%s)"), old_name,
+                      G_mapset());
+            return 1;
+        }
+    }
+    else
+        strcat(sname, old_name);
+    if (G_name_is_fully_qualified(new_name, tname, tmapset)) {
+        if (strcmp(tmapset, G_mapset()) != 0) {
+            G_warning(_("%s is not in the current mapset (%s)"), new_name,
+                      G_mapset());
+            return 1;
+        }
+    }
+    else
+        strcat(tname, new_name);
+
+    smapset = I_find_signature2(type, old_name, tmapset);
+    if (!smapset) {
+        G_warning(_("%s is missing"), old_name);
+        return 1;
+    }
+
+    if (type == SIGFILE_TYPE_SIG)
+        sprintf(selem, "signatures%csig", HOST_DIRSEP);
+    else if (type == SIGFILE_TYPE_SIGSET) {
+        sprintf(selem, "signatures%csigset", HOST_DIRSEP);
+    }
+
+    G_file_name(old_path, selem, sname, tmapset);
+    G_file_name(new_path, selem, tname, tmapset);
+    if (G_rename_file(old_path, new_path) != 0) {
+        G_warning(_("Unable to rename <%s> to <%s>"), old_name, new_name);
+        return 1;
+    }
+    return 0;
+}
+
+/*!
+ * \brief Get list of existing signatures by type
+ *
+ * Fills passed list with fully qualified names of existing signatures.
+ *
+ * If no mapset is passed, all mapsets in the search path are used.
+ * If no signatures are found, returns 0 and list is set to NULL.
+ *
+ * The function will assign memory for the list. It is up to callee to
+ * free the memory of each list item and the list itself.
```

Comment:
Although this memory leak is insignificant as this function is mainly called from GUI where leaking a few bytes is insignificant, I added a I_free_signatures_list() and integrated it into GUI. Besides GRASS philosophy so far always has been "don't free memory, at the exit of module it will be reclaimed by OS anyway".

Reply:
Thanks for fixing it.

> Besides GRASS philosophy so far always has been "don't free memory, at the exit of module it will be reclaimed by OS anyway"

I don't think it is anymore.

Reply:
> > Besides GRASS philosophy so far always has been "don't free memory, at the exit of module it will be reclaimed by OS anyway"
> 
> I don't think it is anymore.

Does this new approach applies to C code only or Python too? Every (indirect) call to G_store needs G_free afterwards (any Python call to libgis any function returning a string needs to free returned string afterwards).

Reply:
Since we have these G_store calls all over the place, it may not be practical to try to achieve perfection with these in new code. However, for clear cases, let's G_free. Freeing stuff is expected by people who actually learned C and tools such as valgrind help everybody to free correctly.

Another thing to consider is that identifying places where G_store needs G_free should be relatively easy or at least easier than fixing some very specific code which lacks the clean up. The specific G_store case can be easily identified because it is calling G_store :-) although the indirect calls definitely make this hard. However, general G_malloc possibly with some nested arrays potentially require much more understanding of the specific case, so for complex code, freeing should be ideally done by the original author.